File: src/nuitrack_bridge/skeleton_tracker.cc
#include "nuitrack_bridge/skeleton_tracker.h"
#include "nuitrack/modules/SkeletonTracker.h"

#include "nuitrack/Nuitrack.h"
#include <stdexcept>
#include <string>

namespace nuitrack_bridge::skeleton_tracker {
    std::shared_ptr<SkeletonTracker> create_skeleton_tracker() {
        try {
            return tdv::nuitrack::SkeletonTracker::create();
        } catch (const tdv::nuitrack::Exception& e) {
            throw std::runtime_error(std::string("Nuitrack SkeletonTracker::create failed: ") + e.what());
        } catch (const std::exception& e) {
            throw std::runtime_error(std::string("Exception during Nuitrack SkeletonTracker::create: ") + e.what());
        } catch (...) {
            throw std::runtime_error("Unknown exception during Nuitrack SkeletonTracker::create");
        }
    }
}

File: src/nuitrack_bridge/skeleton_tracker.rs
#[cxx::bridge(namespace = "nuitrack_bridge::skeleton_tracker")]
pub mod ffi {

    // Opaque C++ types that Rust will interact with via pointers/references.
    #[namespace = "tdv::nuitrack"] // Map to the original Nuitrack namespace
    unsafe extern "C++" {
        type SkeletonTracker;
    }

    unsafe extern "C++" {
        include!("nuitrack_bridge/skeleton_tracker.h");

        #[cxx_name = "create_skeleton_tracker"]
        pub fn skeleton_tracker_create() -> Result<SharedPtr<SkeletonTracker>>;
    }
}

unsafe impl Send for ffi::SkeletonTracker {}
unsafe impl Sync for ffi::SkeletonTracker {}